### PR TITLE
mathtext: Finetuning sup/super block to match TeX reference

### PR DESCRIPTION
**Pull request by zblz**
As mentioned in #4872, I found that sub/super blocks in mathtext have slightly too much space after them, even after @mdboom fix in #4401. As initial reference, this a TeX rendering of a bunch of sub/superscripts (with E and 5 because they are box-like and separation differences are easier to spot):
subsup_test_tex

The current master renders it as follows:
subsup_test_mpl_master

And with the modifications of this PR, it looks like this:
subsup_test_mpl

Summary of changes:

* Additional kerning of -0.1 em after sub/super blocks.
* Superscripts are shifted away from nucleus by 0.15 em when nucleus is slanted.
* When there is both sub and superscript and nucleus is slanted, superscripts are only shifted an aditional 0.15 em (down from >0.20 em)
* Subscripts are shifted down 0.25 x-height instead of 0.5 when there is no superscript.

Regression images coming soon!

### CODE REVIEW

**Comment by WeatherGod** [comment]
Comparing between the Tex and mathtext, It looks like the superscript is
slightly too high as well.

**Comment by zblz** [comment]
Well, it seems that reducing the space after the sub/subscript is not that simple after all: it works well with numbers, but for slanted characters:
subsup_test_mpl_finetune

TeX, of course, handles these beautifully. I'll try to see where the difference in spacing comes from.

@WeatherGod: indeed it is, I'll try to match that as well.

**Comment by zblz** [comment]
Trying to see why ``$a_{ji}$`` was being rendered with too little kerning, I found that the current master renders ``$a_{i}$`` and ``$a_i$`` differently. After much blood and tears in the depths of ``mathtext.py``, I think it is because calling ``.shrink()`` on a ``Char`` object results in different kerning that calling it on a ``Hlist`` object. The groups between ``{ }`` are parsed as ``Hlist`` objects, whereas the characters in ``$a_i$`` are parsed as ``Char``. The current master shrinks and then converts into a Hlist, and I have changed it so that evereything is an ``Hlist`` before shrinking, so there is better consistency. To showcase this, I made a plot with two lines that only differ by (unnecesary) braces:
```
r'$x_kx^py^{p-2} \quad d_i^jb_jc_kd \quad x^j_i E^0 E^0_u$',
r'${x}_{k}{x}^{p}{y}^{p-2} \quad {d}_{i}^{j}{b}_{j}{c}_{k}{d} \quad {x}^{j}_{i}{E}^{0}{E}^0_u$',
```
Current master renders them clearly different:
subsup_test_mpl_master

And with the changes mentioned they are consistent with each other:
subsup_test_mpl_finetune

I have also finetuned the supsuper spacings to match Computer Modern on TeX, but I have found that these do not work properly for other fontsets.... Here is the reference TeX rendering (with text.usetex=True):
subsup_test_mpl_tex

And with this PR:
subsup_test_mpl_finetune

Compared to the current master:
subsup_test_mpl_master

However, not all is rosy, because the same parameters that work for CM fail miserably when using stixsans or stix as mathtext.fontset, so I will try to do a dict that sets them based on the fontset. See for stix (note the superscripts in particular):
subsup_test_mpl_finetune-stix

**Comment by tacaswell** [comment]
@zblz Thank you for taking this on!

**Comment by zblz** [comment]
So I think I've got it under control now for CM, stix and stixsans fontsets. There were a few challenges with integrals (also related to the ``Char`` vs ``Hlist`` issue above), but they look good now and I have added a couple of tests for them. I will amend the baseline images shortly and then the PR can be properly reviewed.

This is how a sample looks with this PR in cm:
subsup_test_mpl_cm_finetune

To be compared with the ``usetex=True`` rendering:
subsup_test_mpl_cm_tex

The same sample also looks good for stix and stixsans:
subsup_test_mpl_stix_finetune
subsup_test_mpl_stixsans_finetune

For reference, this is the rendering with Computer Modern with the current master:
subsup_test_mpl_cm_master

**Comment by zblz** [comment]
Damn, I messed up a rebase. Please stand by...

**Comment by zblz** [comment]
@tacaswell, @mdboom : I think this is now complete, it would be great if you could review it.

Looking at the issue list, this PR will fix the issue with spacing after superscripts reported in #4333.

**Comment by zblz** [comment]
I have also added the sub/supercript parameters for Arev Sans, which is checked for by looking at 
```
rcParams['mathtext.fontset'] == 'custom'
rcParams['mathtext.rm'] == 'sans'
rcParams['font.sans-serif'][0].lower() == 'arev sans'
```
This could be useful for users using it and if #4636 is accepted.

**Comment by zblz** [comment]
I just realised that doing all this I messed up the math accents and they are now shifted to the left:
mathtext_cm_26

However, the tests do not catch this! Probably the accents are too small and don't amount to a high enough RMS. I'll try to find how I messed them up and fix it.

**Comment by zblz** [comment]
So I found that the parameters I used for the images in the last comment above resulted in some conflicts for superscripts in slanted fonts (most notably $f$, which has weird kerning in most fonts). I have improved the superscript placement for this by adding extra kerning based on the height of the last character. It is not TeX-good, but I think TeX is doing optical kerning between the nucleus and the superscripted font, and I don't think we want to go there for a small improvement.

I have only added the png baseline images so the diff is easier to read for review. I added all the baseline images that include sub/superscripts, not only the ones failing (with PR #4887 these sets should be the same, but right now they are not). I will add the pdfs and svgs later, so Travis will fail until then.

**Comment by mdboom** [comment]
Thanks for all this work.  I've been on vacation, but am now back and will review this soon.

File: lib/matplotlib/mathtext.py [hunk 1]
```diff
@@ -2726,48 +2759,93 @@ def subsuper(self, s, loc, toks):
...
+
+        fs = rcParams['mathtext.fontset']
+        # If a custom fontset is used, check if it is Arev Sans, otherwise use
+        # CM parameters.
+        if fs == 'custom':
+            if (rcParams['mathtext.rm'] == 'sans' and
+                    rcParams['font.sans-serif'][0].lower() == 'Arev Sans'.lower()):
+                fs = 'arevsans'
+            else:
+                fs = 'cm'
```
Comment (mdboom) [inline lib/matplotlib/mathtext.py]: For clarity, can you separate this chunk out to its own private function, perhaps called `_get_fontset_name`?

File: lib/matplotlib/mathtext.py [hunk 2]
```diff
@@ -2726,48 +2759,93 @@ def subsuper(self, s, loc, toks):
...
-            shift_down = nucleus.depth + SUBDROP * xHeight
+        # We remove kerning for consistency (otherwise it will compute kerning
+        # based on non-shrinked characters and may put them very close together
+        # when superscripted)
+        # We change the width of the last character to match the advance to
+        # consider some fonts with weird metrics: e.g. stix's f has a width of
+        # 7.75 and a kerning of -4.0 for an advance of 3.72, and we want to put
+        # the superscript at the advance
+        last_char = nucleus
+        if isinstance(nucleus,Hlist):
```
Comment (mdboom) [inline lib/matplotlib/mathtext.py]: Space after comma

File: lib/matplotlib/mathtext.py [hunk 3]
```diff
@@ -2726,48 +2759,93 @@ def subsuper(self, s, loc, toks):
             result = Hlist([vlist])
             return [result]
 
-        # Handle regular sub/superscripts
-        shift_up = nucleus.height - SUBDROP * xHeight
-        if self.is_dropsub(nucleus):
-            shift_down = nucleus.depth + SUBDROP * xHeight
+        # We remove kerning for consistency (otherwise it will compute kerning
```
Comment (mdboom) [inline lib/matplotlib/mathtext.py]: Maybe clarfiy: "We remove the kerning on the last character for consistency" (this doesn't remove the inter-character kerning).

**Comment by mdboom** [comment]
This looks great to me (other than my minor comments, and the need to resolve whatever is going on with the test failures).

**Comment by zblz** [comment]
@mdboom: I implemented your comments, and added all the rest of the baseline images (I didn't want to clutter it with the svgs and pdfs for review initially). I have replaced all baseline images that contain a subsuper, independently of whether they were failing, because current tolerance is not strict enough to fail for the position a single superscript (see #4887).

**Comment by zblz** [comment]
And for reference here, here is some sample text as rendered with 341d481 in computer modern:

subsup_test_mpl_cm_finetune

Reference rendereing with ``usetex=True``:

subsup_test_mpl_tex_finetune

And for stix and stixsans:

subsup_test_mpl_stix_finetune

subsup_test_mpl_stixsans_finetune

**Comment by mdboom** [comment]
Thanks!  I'm happy to merge this once Travis passes.

**Comment by WeatherGod** [comment]
hold on. The second to last line has some spacing issues. The "f" and "2"
and "9" (both as a superscript and as a regular numeral) are much too close.

**Comment by WeatherGod** [comment]
Also, the "prime" symbol in that line is much too close, especially when
the "f" is italic.

**Comment by WeatherGod** [comment]
Also, our mathtext might be parsing something slightly differently than
LaTeX is. In that same line, LaTeX sees the prime symbol as being on its
own and positions it as if it was a normal piece of text (nice and big),
but everywhere else, it is some sort of weird, scraggly looking superscript.

**Comment by zblz** [comment]
@WeatherGod: Yes, the kerning for the f is particularly tight, but it is tight everywhere, as you mention for the regular f and numerals 2 and 9 (and that's why I added them to that line), so it is not an issue with the sub/super script. I tried to minimize it for cm, where it looks pretty good, but did not manage for stix of stixsans. Maybe the kerning for the f should be slightly widened elsewhere to match TeX.

The prime issue only happens for stix and stixsans, where it seems to match $\prime$ to an already superscripted prime, as opposed to a nice and big prime for TeX and cm. This I also believe is outside the scope of this PR, but should definitely be fixed so they are consistent.

**Comment by zblz** [comment]
I fixed a test failure in ``test_axes`` that came from having no nucleus in mathtext (e.g., 'a$^2$'), so I added a test in mathtext so it can be caught by running only the ``test_mathtext`` tests. There were also a couple of superscripts in other tests that were failing because I had not replaced the baseline images. Should be fine now.

**Comment by mdboom** [comment]
I agree that the stix issues with the prime symbol and some of the other miscellaneous spacing is a  separate issue from this PR, and indeed this PR makes things no worse than it was before in that regard.

**Comment by WeatherGod** [comment]
Ok, well, let's create some issue tickets for those so that we don't forget to fix them. This is great work @zblz!

**Comment by WeatherGod** [comment]
Ok, I created issue #4987 to track the issues I noted above. Great work!

Mike, I have not reviewed a single line of code here (I wouldn't know what any of it means anyway). So, I'll leave merging this up to you?

**Pull request merged**